### ISSUE TITLE
Indexing multidimensional structured masked array fails to translate fill_value correctly, leading to broadcasting errors ValueError when calling .filled()

### ISSUE THREAD

**Issue by gerritholl** [numpy/numpy#6723]
When using a structured masked array where one of the structure elements has a multidimensional dtype, indexing does not correctly set the `fill_value` attribute for the new masked array.  This leads to a `fill_value` with a shape that is incompatible with `.data` or `.mask`, which inevitably leads to broadcasting problems when `fill_value` is used, such as when calling `filled()`:

```
In [332]: A = ma.masked_array(data=[([0,1,2],), ([3,4,5],)], mask=[([True, False, False],), ([False, True, False],)], dtype=[("A", ">i2", (3,))])

In [339]: A["A"][:, 0].filled()
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-339-ddde509d73bf> in <module>()
----> 1 A["A"][:, 0].filled()

/home/users/gholl/venv/stable-3.5/lib/python3.5/site-packages/numpy/ma/core.py in filled(self, fill_value)
   3573             result = self._data.copy('K')
   3574             try:
-> 3575                 np.copyto(result, fill_value, where=m)
   3576             except (TypeError, AttributeError):
   3577                 fill_value = narray(fill_value, dtype=object)

ValueError: could not broadcast input array from shape (3) into shape (2)

In [341]: A.shape, A.fill_value.shape, A["A"].shape, A["A"].fill_value.shape, A["A"][:, 0].shape, A["A"][:, 0].fill_value.shape
Out[341]: ((2,), (), (2, 3), (3,), (2,), (3,))
```

Indeed, the fill value doesn't change upon indexing the masked array — only its dtype does:

```
In [351]: A.fill_value, A["A"].fill_value, A["A"][:, 0].fill_value
Out[351]: 
(([16959, 16959, 16959],),
 array([16959, 16959, 16959], dtype=int16),
 array([16959, 16959, 16959], dtype=int16))
```

**Comment by gerritholl** [comment]
Would it be sensible to require that a `fill_value` should always have `.ndim==0`?

**Comment by gerritholl** [comment]
A more commonly used method through to which this propagates is `pcolor` — I found this bug through matplotlibs `pcolor` function/method, and I am currently using a workaround (explicitly setting `X._fill_value = X.fill_value.flat[0]`) so that `pcolor` works.

**Comment by gerritholl** [comment]
@charris  The reason dimensionality of the `fill_value` goes up upon indexing is because we access a multidimensional field, `A["A"]` in the above example.  `A.dtype == A.fill_value.dtype == dtype([('A', '>i2', (3,))])`, `A.shape == (2,)`, `A.fill_value.dtype == ()`.  However, `A["A"].dtype == dtype('>i2')` and `A["A"].shape == (2, 3)`.  So any time we index a field that is multidimensional, `A[field]` will have higher dimensionality than `A`.  In this example, `A.ndim == 1` and `A["A"].ndim == 2`, `A.fill_value.ndim == 0`, and finally `A["A"].fill_value.ndim == 1`.  It's this last dimensionality that causes problems.

**Comment by gerritholl** [comment]
More compactly put: `M[field].ndim == M.ndim + len(M.dtype[field].shape)`

**Comment by eric-wieser** [comment]
This was addressed in 090e85e7bd77e67c8b6bac63fc4b22331eac1950

**Comment by mhvk** [comment]
This has been fixed (I assume by #6728), so closing.
